master/buildbot/data/buildrequests.py

```diff
@@ -98,6 +98,24 @@ def startConsuming(self, callback, options, kwargs):
         return self.master.mq.startConsuming(callback,
                                              ('buildrequests', None, None, None, None))
 
+    @defer.inlineCallbacks
+    def control(self, method, kwargs):
+        # first, try to claim the request; if this fails, then it's too late to
+        # cancel the build anyway
+        try:
+            yield self.master.data.updates.claimBuildRequests([self.id])
+        except buildrequests.AlreadyClaimedError:
+            builds = yield self.master.data.get(("buildrequests", brid, "builds"))
```

Comment:
need to add a comment about it, as I dont have any suggestion on how to manage this race condition.